### PR TITLE
Setting up experiment MARK01

### PR DESCRIPTION
**Pull request by gordonkoehn**
Implementing the experimental workflow of 
- #73 

Produces this massive DAG:
[mark01_dag.pdf](https://github.com/cbg-ethz/PYggdrasil/files/11881828/mark01_dag.pdf)

This PR closes:
- #91 
- #92 

Before we can run the experiment it is worth investigating why the scyphylo AD distance still fails sometimes. 
- #89

### CODE REVIEW

File: src/pyggdrasil/serialize/_to_json.py [hunk 1]
```diff
@@ -207,9 +212,12 @@ def save_metric_result(iteration: list[int], result: list[float], out_fp: Path)
...
+def read_metric_result(fullpath: Path) -> tuple[list[Union[int, str]], list[float]]:
     """Reads in all metric results from JSON file for a given run."""
     with open(fullpath, "r") as f:
         metric_dict = json.load(f)
 
-    return metric_dict["iteration"], metric_dict["result"]
+    # get axis name
+    axis_name = list(metric_dict.keys())[0]
+
+    return metric_dict[axis_name], metric_dict["result"]
```
Comment (gordonkoehn) [inline src/pyggdrasil/serialize/_to_json.py:223]: Reusing these functions - made them more flexible.

File: src/pyggdrasil/tree.py [hunk 2]
```diff
@@ -56,5 +56,21 @@ def print_topo(self) -> None:
...
+        Args:
+              node: AnyTree node to be converted
+        Returns:
+            TreeNode with the same topology as the AnyTree node
+        """
+        children = [
+            TreeNode.convert_anytree_to_treenode(child) for child in node.children
+        ]
+        return TreeNode(name=node.name, children=children)
+
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree.py:74]: Tree Conversions still haunt me :) 

Hope this was the last one.

File: src/pyggdrasil/tree_inference/_file_id.py [hunk 3]
```diff
@@ -208,15 +208,25 @@ def from_str(cls, str_id: str):
         """
         # split string by underscore and assign to attributes
         # CS_1-T_d_10-100_0.01_0.01_0.01_true_UXR
-        cs_par1, tree_id, cs_part2 = str_id.split("-")
+        parts = str_id.split("-")
+        cs_part1 = parts[0]
+        tree_id = parts[1]
+        print(tree_id)
+        cs_part2 = "-".join(parts[2:])
+
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_file_id.py:216]: Allowing for scientific notation error rates. 

The trouble with string matching.

File: workflows/mark01.smk [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+# Homozygous mutations [f: False / t: True]
+observe_homozygous = "f" # <-- configure whether to observe homozygous mutations here
+
+# cell attachment strategy [UXR: Uniform Exclude Root / UIR: Uniform Include Root]
+cell_attachment_strategy = "UXR" # <-- configure cell attachment strategy here
+
+#####################
+# True Tree Parameters
+tree_types = ["r"] # <-- configure tree type here ["r","s","d"]
+tree_seeds = [42] # <-- configure tree seed here
```
Comment (gordonkoehn) [inline workflows/mark01.smk:44]: Easy to choose more than one true tree.

File: workflows/mark01.smk [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    run:
+        import pyggdrasil as yg
+        from pathlib import Path
+        import matplotlib.pyplot as plt
+        # load the distances
+        tree_ids, distances =  yg.serialize.read_metric_result(Path(input.distances))
+        # make the histogram
+        fig, axs = plt.subplots(1,1,tight_layout=True)
+        # We can set the number of bins with the *bins* keyword argument.
+        axs.hist(distances, bins='auto')
```
Comment (gordonkoehn) [inline workflows/mark01.smk:94]: A preference for binning? Currently using the "auto = "`Maximum of the ‘sturges’ and ‘fd’ estimators. `

Comment (gordonkoehn) [reply workflows/mark01.smk:94]: https://numpy.org/doc/stable/reference/generated/numpy.histogram_bin_edges.html#numpy.histogram_bin_edges

**Review by gordonkoehn — COMMENTED** [review]
Full MARK01 workflow - fix AD distance next. 

So far I have not seen any interesting histograms during testing. 

Huntress either gave a tree of MP3 = 0 or 1 for all setups. 

Hope more exciting is to come.

File: workflows/mark01.smk (hunk 5 above)
Comment (pawel-czyz) [reply workflows/mark01.smk:94]: Any choice is good, I think – we'll probably want to load the data and play with the bin settings in a notebook anyway (or look at other visualisations in more exploratory manner)

File: tests/tree_inference/test_file_id.py [hunk 6]
```diff
@@ -105,3 +122,26 @@ def test_cell_simulation_id_from_str(cell_simulation_id) -> None:
     assert csi.na_rate == cell_simulation_id.na_rate
     assert csi.observe_homozygous == cell_simulation_id.observe_homozygous
     assert csi.strategy == cell_simulation_id.strategy
+
+
+def test_cell_simulation_id_from_str_scientific_notation(
```
Comment (pawel-czyz) [inline tests/tree_inference/test_file_id.py:127]: Nice!

**Review by pawel-czyz — APPROVED** [review]
LGTM!

File: src/pyggdrasil/tree_inference/_file_id.py (hunk 3 above)
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_file_id.py:216]: Yes, I completely see how one could do this now! Understand Snakemake better now. Would do things differently now :)  Would have made the output file names more human-readable. 

Would be quite a bit of refactoring in `workflows`, `tests` and `tree_inference` to change this now.

I don't see any danger of the error rates representation as numbers breaking down now. 

I'll stick to the current solution. This is probably something I won't change after running the first experiments.

If final output files get tiring to read I write a rule to rename files for ease of finding the files one wants.

**Pull request merged**